### PR TITLE
malus - mockable overseer mvp

### PR DESCRIPTION
**Pull request by drahnr**
As described in #3205 implements the MVP variant w/o going all the way into proc macro land just yet (see #2949 for that).

* [x] add simple interception infra and a demo malicious subsystem dropping every second message
* [x] create `[[bin]]` targets for malicious nodes

Obsoletes #2696

### CODE REVIEW

File: cli/src/command.rs [hunk 1]
```diff
@@ -181,53 +181,64 @@ fn ensure_dev(spec: &Box<dyn service::ChainSpec>) -> std::result::Result<(), Str
 	}
 }
 
-/// Parses polkadot specific CLI arguments and run the service.
-pub fn run() -> Result<()> {
-	let cli = Cli::from_args();
+#[cfg(feature = "malus")]
+/// Launch a node, accepting arguments just like a regular node,
+/// accepts an alternative overseer generator, to adjust behavior
+/// for integration tests as needed.
```
Comment (ordian) [inline cli/src/command.rs:187]:
```suggestion
/// Launch a node, accepting arguments just like a regular node,
/// accepts an alternative overseer generator, to adjust behavior
/// for integration tests as needed.
#[cfg(feature = "malus")]
```

File: node/service/src/overseer.rs [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+	Error,
+	Registry,
+	IsCollator,
+	Block,
+	SpawnNamed,
+	Hash,
+	AuthorityDiscoveryApi,
+};
+use std::sync::Arc;
+use	polkadot_network_bridge::RequestMultiplexer;
```
Comment (ordian) [inline node/service/src/overseer.rs:27]: mixed tabs and spaces after `use`

File: node/service/src/overseer.rs [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+use polkadot_node_collation_generation::CollationGenerationSubsystem;
+use polkadot_collator_protocol::{CollatorProtocolSubsystem, ProtocolSide};
+use polkadot_network_bridge::NetworkBridge as NetworkBridgeSubsystem;
+use polkadot_node_core_provisioner::ProvisioningSubsystem as ProvisionerSubsystem;
+use polkadot_node_core_runtime_api::RuntimeApiSubsystem;
+use polkadot_statement_distribution::StatementDistribution as StatementDistributionSubsystem;
+use polkadot_availability_recovery::AvailabilityRecoverySubsystem;
+use polkadot_approval_distribution::ApprovalDistribution as ApprovalDistributionSubsystem;
+use polkadot_node_core_approval_voting::ApprovalVotingSubsystem;
+use polkadot_gossip_support::GossipSupport as GossipSupportSubsystem;
```
Comment (ordian) [inline node/service/src/overseer.rs:79]: could you put all imports above `pub struct`

File: node/service/src/overseer.rs [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+			keystore.clone(),
+		),
+	};
+	Ok(all_subsystems)
+}
+
+
+/// Trait for the fn generating the overseer.
+///
+/// Default behavior is to creat an unmodified overseer, as `RealOverseerGen`
```
Comment (ordian) [inline node/service/src/overseer.rs:222]:
```suggestion
/// Default behavior is to create an unmodified overseer, as `RealOverseerGen`
```

**Review by ordian — APPROVED** [review]
This looks cool, just some formatting nits.

File: node/service/src/overseer.rs [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+pub use polkadot_node_core_approval_voting::ApprovalVotingSubsystem;
+pub use polkadot_gossip_support::GossipSupport as GossipSupportSubsystem;
+
+/// Arguments passed for overseer construction.
+pub struct OverseerGenArgs<'a, Spawner, RuntimeClient> where
+	RuntimeClient: 'static + ProvideRuntimeApi<Block> + HeaderBackend<Block> + AuxStore,
+	RuntimeClient::Api: ParachainHost<Block> + BabeApi<Block> + AuthorityDiscoveryApi<Block>,
+	Spawner: 'static + SpawnNamed + Clone + Unpin,
+{
+	pub leaves: Vec<BlockInfo>,
```
Comment (rphmeier) [inline node/service/src/overseer.rs:64]: Docs on these would be nice

File: node/service/src/overseer.rs [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+	where
+		RuntimeClient: 'static + ProvideRuntimeApi<Block> + HeaderBackend<Block> + AuxStore,
+		RuntimeClient::Api: ParachainHost<Block> + BabeApi<Block> + AuthorityDiscoveryApi<Block>,
+		Spawner: 'static + SpawnNamed + Clone + Unpin {
+		let gen = RealOverseerGen;
+		RealOverseerGen::generate::<Spawner, RuntimeClient>(&gen, args)
+	}
+	// It would be nice to make `create_subsystems` part of this trait,
+	// but the amount of generic arguments that would be required as
+	// as consequence make this rather annoying to implement and use.
```
Comment (rphmeier) [inline node/service/src/overseer.rs:233]: 👍

**Comment by drahnr** [comment]
Current master (`40d1612b83`) : 5min 47s
PR: 5min 50s

so my initial feeling of things being extra slow was wrong, just limitations of the laptop

**Comment by drahnr** [comment]
bot merge

**Comment by drahnr** [comment]
bot merge abort

**Comment by drahnr** [comment]
bot merge cancel

**Pull request merged**